# Project Plan: Sales Data Analysis Using Python + SQL

## Purpose

This project aims to analyze sales data to uncover trends, optimize inventory management, and evaluate marketing effectiveness. By leveraging Python libraries such as Pandas, Matplotlib, and NumPy, I will explore patterns in order to provide actionable business insights.

## Methods

### **1. Data Cleaning & Preprocessing:**

- Load and inspect the datasets using Pandas.
- Handle missing values and inconsistent data entries.
- Convert categorical data into appropriate formats.
- Merge relevant datasets to create a unified view for analysis.

### **2. Exploratory Data Analysis (EDA):**

- Generate summary statistics for key numerical and categorical variables.
- Analyze sales distribution and seasonal trends.
- Identify top-selling and least-selling products.
- Investigate correlations between pricing, sales volume, and marketing efforts.

### **3. Data Visualization:**

- Create bar charts and histograms to show sales trends across different categories.
- Develop time-series plots to analyze seasonal patterns in sales.
- Use heatmaps to identify relationships between different features (e.g., sales and marketing spend).
- Visualize inventory turnover and stockout trends.

### **4. Research Questions:**

1. What are the top-selling products and categories based on sales revenue and order volume?
2. How does pricing impact sales performance?
3. What is the relationship between marketing spend and sales growth?
4. How efficient is the delivery process, and what factors influence delivery times?
5. Are there seasonal patterns in sales, and how do they affect inventory levels?
6. How can inventory management be optimized to reduce stockouts and overstocking?

### **5. Tools & Technologies:**

- **Python Libraries:** Pandas, NumPy, Matplotlib, etc.
- **Jupyter Notebook:** For code execution and analysis
- **SQL (?):** Querying structured datasets for deeper insights

## Expected Outcomes

- A well-structured Jupyter Notebook containing data analysis and visualizations.
- Insights into product sales trends, customer purchasing behaviors, and marketing effectiveness.
- Data-driven recommendations for improving inventory management and delivery performance.
- Well-documented Python and SQL? scripts for reproducibility and future analysis.
- (?) Tableau Visuals



## Documentation Overview

This dataset consists of 9 different CSV files containing different categories of information. For this initial analysis using Python, I will keep the files separate. I will leverage SQL to manipulate the tables and produce further insights later on. This document is organized by file, where the EDA process is repeated for each one. I will clean and explore each file individually before laying out all of my findings at the end to produce a report containing key business insights and recommendations. 

### File 1: Blinket Products
This file contains key product information. Upon viewing it, I developed these questions to answer:
- What is the average list price and profit margin across all products?
- Which products have the highest and lowest profit margin?
- fill

\
*Note: the 'mrp' column stands for Maximum Retail Price.*

In [110]:
# importing libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# loading csv into a dataframe
product_df = pd.read_csv('/Users/tovi/Documents/blinkit_products.csv')
product_df

product_id   product_name             category                    brand  \
0        153019         Onions  Fruits & Vegetables               Aurora LLC   
1         11422       Potatoes  Fruits & Vegetables           Ramaswamy-Tata   
2        669378       Potatoes  Fruits & Vegetables          Chadha and Sons   
3        848226       Tomatoes  Fruits & Vegetables           Barad and Sons   
4        890623         Onions  Fruits & Vegetables  Sangha, Nagar and Varty   
..          ...            ...                  ...                      ...   
263      444361  Pain Reliever             Pharmacy   Prakash, Bawa and Kale   
264      679284    Cough Syrup             Pharmacy                 Pant LLC   
265      240179    Cough Syrup             Pharmacy                 Ram-Suri   
266      673058    Cough Syrup             Pharmacy              Balan-Madan   
267      114414       Vitamins             Pharmacy               Kara-Golla   

      price      mrp  margin_percentage  shelf_life_days  min_stock_level  \
0    947.95  1263.93               25.0                3               13   
1    127.16   169.55               25.0                3               20   
2    212.14   282.85               25.0                3               23   
3    209.59   279.45               25.0                3               10   
4    354.52   472.69               25.0                3               27   
..      ...      ...                ...              ...              ...   
263  822.63  1028.29               20.0              365               20   
264  877.89  1097.36               20.0              365               28   
265   90.56   113.20               20.0              365               20   
266  765.76   957.20               20.0              365               30   
267  832.06  1040.07               20.0              365               26   

     max_stock_level  
0                 88  
1                 65  
2                 70  
3                 51  
4                 55  
..               ...  
263               71  
264               95  
265               56  
266               94  
267               66  

[268 rows x 10 columns]

In [80]:
# general descriptive stats
product_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 268 entries, 0 to 267
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   product_id         268 non-null    int64  
 1   product_name       268 non-null    object 
 2   category           268 non-null    object 
 3   brand              268 non-null    object 
 4   price              268 non-null    float64
 5   mrp                268 non-null    float64
 6   margin_percentage  268 non-null    float64
 7   shelf_life_days    268 non-null    int64  
 8   min_stock_level    268 non-null    int64  
 9   max_stock_level    268 non-null    int64  
dtypes: float64(3), int64(4), object(3)
memory usage: 21.1+ KB


In [57]:
# checking for null values
null_check = product_df.isnull().sum()
print(null_check)

product_id           0
product_name         0
category             0
brand                0
price                0
mrp                  0
margin_percentage    0
shelf_life_days      0
min_stock_level      0
max_stock_level      0
dtype: int64


All good, now this file is ready for analysis! First let's take a look at some averages:

In [93]:
product_df[['price', 'margin_percentage']].mean()

price                488.356828
margin_percentage     27.779851
dtype: float64

The average product list price is $488.36, and the average profit margin across all products is 27.78%. 

### File 2: Blinket Deliveries
This file contains key information on delivery times.\
Research Questions:
- How frequently are items not delivered on time?
- Are any delivery partners, customers, or stores having repeated issues with delivery?

In [62]:
# loading csv into a dataframe
delivery_df = pd.read_csv('/Users/tovi/Documents/blinkit_orders.csv')
delivery_df.head()

order_id  customer_id           order_date promised_delivery_time  \
0  1961864118     30065862  2024-07-17 08:34:01    2024-07-17 08:52:01   
1  1549769649      9573071  2024-05-28 13:14:29    2024-05-28 13:25:29   
2  9185164487     45477575  2024-09-23 13:07:12    2024-09-23 13:25:12   
3  9644738826     88067569  2023-11-24 16:16:56    2023-11-24 16:34:56   
4  5427684290     83298567  2023-11-20 05:00:39    2023-11-20 05:17:39   

  actual_delivery_time delivery_status  order_total payment_method  \
0  2024-07-17 08:47:01         On Time      3197.07           Cash   
1  2024-05-28 13:27:29         On Time       976.55           Cash   
2  2024-09-23 13:29:12         On Time       839.05            UPI   
3  2023-11-24 16:33:56         On Time       440.23           Card   
4  2023-11-20 05:18:39         On Time      2526.68           Cash   

   delivery_partner_id  store_id  
0                63230      4771  
1                14983      7534  
2                39859      9886  
3                61497      7917  
4                84315      2741

In [105]:
# general descriptive stats
product_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 268 entries, 0 to 267
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   product_id         268 non-null    int64  
 1   product_name       268 non-null    object 
 2   category           268 non-null    object 
 3   brand              268 non-null    object 
 4   price              268 non-null    float64
 5   mrp                268 non-null    float64
 6   margin_percentage  268 non-null    float64
 7   shelf_life_days    268 non-null    int64  
 8   min_stock_level    268 non-null    int64  
 9   max_stock_level    268 non-null    int64  
dtypes: float64(3), int64(4), object(3)
memory usage: 21.1+ KB


In [24]:
# checking for null values
null_check = delivery_df.isnull().sum()
print(null_check)

order_id                  0
customer_id               0
order_date                0
promised_delivery_time    0
actual_delivery_time      0
delivery_status           0
order_total               0
payment_method            0
delivery_partner_id       0
store_id                  0
dtype: int64


### File 3: Blinket Units
The third file contains information on unit sales and pricing:

In [32]:
df3 = pd.read_csv('/Users/tovi/Documents/blinkit_order_items.csv')
df3

order_id  product_id  quantity  unit_price
0     1961864118      642612         3      517.03
1     1549769649      378676         1      881.42
2     9185164487      741341         2      923.84
3     9644738826      561860         1      874.78
4     5427684290      602241         2      976.55
...          ...         ...       ...         ...
4995  1669690997      925482         3      475.04
4996  8340761903      124290         2       39.65
4997  5936301790      491314         3      973.44
4998  5710579377      319388         2      657.80
4999  2494813730      136533         1      826.21

[5000 rows x 4 columns]

The fourth file:

In [40]:
df4 = pd.read_csv('/Users/tovi/Documents/blinkit_marketing_performance.csv')
df4

campaign_id          campaign_name        date target_audience  \
0          548299      New User Discount  2024-11-05         Premium   
1          390914        Weekend Special  2024-11-05        Inactive   
2          834385         Festival Offer  2024-11-05        Inactive   
3          241523             Flash Sale  2024-11-05        Inactive   
4          595111       Membership Drive  2024-11-05       New Users   
...           ...                    ...         ...             ...   
5395       381623       Membership Drive  2023-03-17       New Users   
5396       493317     Category Promotion  2023-03-17         Premium   
5397       879292  App Push Notification  2023-03-17             All   
5398       461129         Email Campaign  2023-03-17       New Users   
5399       282514       Referral Program  2023-03-17       New Users   

           channel  impressions  clicks  conversions    spend  \
0              App         3130     163           78  1431.85   
1              App         3925     494           45  4506.34   
2            Email         7012     370           78  4524.23   
3              SMS         1115     579           86  3622.79   
4            Email         7172     795           54  2888.99   
...            ...          ...     ...          ...      ...   
5395         Email         8331     809           92  2363.07   
5396  Social Media         5216     860           34  3922.53   
5397           App         8480     734           20  1990.75   
5398         Email         3569     568           10  4246.47   
5399  Social Media         5967     439           78  1511.93   

      revenue_generated  roas  
0               4777.75  3.60  
1               6238.11  2.98  
2               2621.00  2.95  
3               2955.00  2.84  
4               8951.81  2.22  
...                 ...   ...  
5395            3710.79  1.64  
5396            7701.25  2.01  
5397            6650.41  3.98  
5398            5880.42  2.50  
5399            8322.01  3.08  

[5400 rows x 11 columns]

The fifth file: **NOTE -NEW version, or og version**

In [52]:
df5 = pd.read_csv('/Users/tovi/Documents/blinkit_inventoryNew.csv')
df5

product_id    date  stock_received  damaged_stock
0          153019  Mar-23               4              1
1          848226  Mar-23               4              1
2          965755  Mar-23               1              0
3           39154  Mar-23               4              0
4           34186  Mar-23               3              1
...           ...     ...             ...            ...
18100      837480  Nov-24               3              0
18101       56589  Nov-24               3              0
18102      679284  Nov-24               3              0
18103      673058  Nov-24               3              0
18104      114414  Nov-24               3              0

[18105 rows x 4 columns]

The sixth file: **NOTE**

In [49]:
df6 = pd.read_csv('/Users/tovi/Documents/blinkit_inventory.csv')
df6

product_id        date  stock_received  damaged_stock
0          153019  17-03-2023               4              2
1          848226  17-03-2023               4              2
2          965755  17-03-2023               1              0
3           39154  17-03-2023               4              0
4           34186  17-03-2023               3              2
...           ...         ...             ...            ...
75167      298985  05-11-2024               3              0
75168      679284  05-11-2024               3              0
75169      240179  05-11-2024               3              0
75170      673058  05-11-2024               3              2
75171      114414  05-11-2024               3              0

[75172 rows x 4 columns]

The seventh file: **NOTE maybe to be excluded**

In [55]:
df7 = pd.read_csv('/Users/tovi/Documents/blinkit_delivery_performance.csv')
df7

order_id  delivery_partner_id        promised_time  \
0     1961864118                63230  2024-07-17 08:52:01   
1     1549769649                14983  2024-05-28 13:25:29   
2     9185164487                39859  2024-09-23 13:25:12   
3     9644738826                61497  2023-11-24 16:34:56   
4     5427684290                84315  2023-11-20 05:17:39   
...          ...                  ...                  ...   
4995  1669690997                90914  2023-12-25 16:05:20   
4996  8340761903                27952  2023-11-27 09:38:43   
4997  5936301790                 9590  2024-06-21 19:23:09   
4998  5710579377                29940  2024-06-06 15:12:13   
4999  2494813730                32851  2023-08-23 12:20:18   

              actual_time  delivery_time_minutes  distance_km delivery_status  \
0     2024-07-17 08:47:01                   -5.0         0.96         On Time   
1     2024-05-28 13:27:29                    2.0         0.98         On Time   
2     2024-09-23 13:29:12                    4.0         3.83         On Time   
3     2023-11-24 16:33:56                   -1.0         2.76         On Time   
4     2023-11-20 05:18:39                    1.0         2.63         On Time   
...                   ...                    ...          ...             ...   
4995  2023-12-25 16:10:20                    5.0         1.86         On Time   
4996  2023-11-27 09:36:43                   -2.0         3.31         On Time   
4997  2024-06-21 19:26:09                    3.0         2.44         On Time   
4998  2024-06-06 15:10:13                   -2.0         1.76         On Time   
4999  2023-08-23 12:21:18                    1.0         2.70         On Time   

     reasons_if_delayed  
0                   NaN  
1               Traffic  
2               Traffic  
3                   NaN  
4               Traffic  
...                 ...  
4995            Traffic  
4996                NaN  
4997            Traffic  
4998                NaN  
4999            Traffic  

[5000 rows x 8 columns]

The eigth file:

In [59]:
df8 = pd.read_csv('/Users/tovi/Documents/blinkit_customers.csv')
df8

customer_id      customer_name                        email  \
0        97475543      Niharika Nagi       ektataneja@example.org   
1        22077605       Megha Sachar         vedant45@example.com   
2        47822591         Hema Bahri        samiazaan@example.com   
3        79726146         Zaitra Vig        ishanvi87@example.org   
4        57102800       Januja Verma     atideshpande@example.org   
...           ...                ...                          ...   
2495     48002829       Daksh Mandal   varkeymohammed@example.com   
2496     57392064       Lavanya Jain      deshpandeom@example.com   
2497     71688530         Umang Dash  gokhalenicholas@example.org   
2498     89051578    Zinal Natarajan        vyasfrado@example.com   
2499     78738569  Eiravati Sundaram          onkar24@example.com   

             phone                                       address  \
0     912987579691           23, Nayar Path, Bihar Sharif-154625   
1     915123179717           51/302, Buch Chowk\nSrinagar-570271   
2     910034076149            941\nAnne Street, Darbhanga 186125   
3     916264232390                43/94, Ghosh, Alappuzha 635655   
4     917293526596                      06\nOm, Ambarnath 477463   
...            ...                                           ...   
2495  919587731286   28/42, Venkataraman Ganj, Kishanganj-360157   
2496  916137420258                 391, Edwin Path, Korba 202726   
2497  917595275963                 151, Parmer Ganj\nLoni 906431   
2498  911994009448  H.No. 172, Khanna Street, Pudukkottai 610564   
2499  914320353753             H.No. 791, Sur Path\nBally 982559   

              area  pincode registration_date customer_segment  total_orders  \
0            Udupi   321865        2023-05-13          Premium            13   
1          Aligarh   149394        2024-06-18         Inactive             4   
2        Begusarai   621411        2024-09-25          Regular            17   
3        Kozhikode   826054        2023-10-04              New             4   
4     Ichalkaranji   730539        2024-03-22         Inactive            14   
...            ...      ...               ...              ...           ...   
2495        Mumbai    45238        2024-01-25         Inactive            17   
2496         Udupi   688100        2024-06-23          Regular             4   
2497        Kavali   528749        2024-03-18          Regular             1   
2498         Alwar   586734        2024-09-06          Premium            12   
2499         Dewas   342448        2023-05-10         Inactive             4   

      avg_order_value  
0              451.92  
1              825.48  
2             1969.81  
3              220.09  
4              578.14  
...               ...  
2495           754.33  
2496          1540.81  
2497          1541.22  
2498          1185.50  
2499           376.50  

[2500 rows x 11 columns]

The ninth file:

In [62]:
df9 = pd.read_csv('/Users/tovi/Documents/blinkit_customer_feedback.csv')
df9

feedback_id    order_id  customer_id  rating  \
0         2234710  1961864118     30065862       4   
1         5450964  1549769649      9573071       3   
2          482108  9185164487     45477575       3   
3         4823104  9644738826     88067569       4   
4         3537464  5427684290     83298567       3   
...           ...         ...          ...     ...   
4995      9024060  1669690997     62600289       3   
4996      4715968  8340761903     53640286       3   
4997      9621021  5936301790     87059497       4   
4998      1134095  5710579377     67310893       4   
4999      6578433  2494813730     28663279       4   

                             feedback_text feedback_category sentiment  \
0            It was okay, nothing special.          Delivery   Neutral   
1                 The order was incorrect.    App Experience  Negative   
2            It was okay, nothing special.    App Experience   Neutral   
3         The product met my expectations.    App Experience   Neutral   
4     Product was damaged during delivery.          Delivery  Negative   
...                                    ...               ...       ...   
4995            Taste was not as expected.   Product Quality  Negative   
4996    Average experience, could improve.   Product Quality   Neutral   
4997    Average experience, could improve.    App Experience   Neutral   
4998    Average experience, could improve.          Delivery   Neutral   
4999         It was okay, nothing special.  Customer Service   Neutral   

     feedback_date  
0       2024-07-17  
1       2024-05-28  
2       2024-09-23  
3       2023-11-24  
4       2023-11-20  
...            ...  
4995    2023-12-25  
4996    2023-11-27  
4997    2024-06-21  
4998    2024-06-06  
4999    2023-08-23  

[5000 rows x 8 columns]

**TO DO**
- create separate doc for analysis w SQL documentation
- upload CSVs to pgadmin
- select files for PYTHON EDA, NOT SQL, reduce it, upload, view, and choose- explain in beginning why certain python files were left out of the python analysis
- should anything be combined?
- choose top 2-3 files for matplot and numpy demo
- Tableau dashboard

## Exploratory 